### PR TITLE
protect keys from writing/deleting/manipulating

### LINKED ISSUE #557: weldxfile: consider removing asdf metadata from user dictionary like operations.
During writing examples (doctests) for WeldxFile, I noticed that it would be bad, if users are able to delete basic asdf meta data like history, software versions etc.
So I'd suggest to remove these data from the users view, restricting it to the data he or she has actually wanted to be in the file (dictionary). Of course the data is still there and will be written as such, but just became immutable. For `history` we already have an interface. So we'd need to another getter for software.

### PR DESCRIPTION
**Pull request by marscher**
## Changes

Protect some keys from reading/writing/deleting/manipulating them. There is a list of attributes in file.py, which are being passed to the protected key dictionary. The WeldxFile now derives from it.
We need to deprecate the `data` attribute, as it allowed raw access.

## Related Issues

Closes #557

## Checks

- [x] updated CHANGELOG.rst
- [x] updated tests
- [x] update example/tutorial notebooks

### CODE REVIEW

**Comment by marscher** [comment]
does this "arguments differ" warnings raised by codacy origin from pylint? I cannot figure out which tool they use. The argument has been renamed from the top-level typing interface named "__m" to "mapping" for better user-experience. I can also add an ignore for this.
Another possibility which might have triggered this is the different (more sane) type annotation. I would also refuse to change that, because it makes more sense to require a key to be hashable in contrast to the super broad TypeVar("_K").

**Comment by CagtayFabry** [comment]
it looks like the pylint run but I am fine with this kind of error (or adding a simple ignore 👍 )

**Comment by vhirtham** [comment]
If you expand the error box, the second line is: 

~~~
Reported by Pylint (Python 3) Time to fix: 5 minutes
~~~

So as @CagtayFabry  said, its pylint. Just add `# pylint: disable=W0221` if you think we can ignore the issue

**Review by CagtayFabry — APPROVED** [review]
great, thanks! 🚀

File: weldx/asdf/file.py [hunk 1]
```diff
@@ -59,9 +70,15 @@ def reset_file_position(fh: SupportsFileReadWrite):
 DEFAULT_ARRAY_COPYING = True
 """Stored Arrays will be copied to memory, or not. If False, use memory mapping."""
 
+_PROTECTED_KEYS = (
+    "history",
+    "asdf_library",
+)
+"""These keys are not not seen, nor can they be manipulated."""
```
Comment (vhirtham) [inline weldx/asdf/file.py:77]: > nor can they be manipulated

Would you bet on this? 😄

File: weldx/asdf/util.py [hunk 2]
```diff
@@ -542,6 +553,80 @@ def _get_instance_shape(
...
+            }
+
+        self._data.update(_mapping)
+
+    def popitem(self) -> Tuple[Hashable, Any]:
+        for k in self.keys():
+            if k in self.protected_keys:
+                continue
+
+            return k, self.pop(k)
```
Comment (vhirtham) [inline weldx/asdf/util.py:609]:
```suggestion
        for k in self.keys():
            if k not in self.protected_keys:
                return k, self.pop(k)
```

**Review by vhirtham — APPROVED** [review]

File: weldx/asdf/file.py (hunk 1 above)
Comment (marscher) [reply weldx/asdf/file.py:77]: Not if you suggest to make a bet :) What have you figured?

File: weldx/asdf/util.py (hunk 2 above)
Comment (marscher) [reply weldx/asdf/util.py:609]: I thought I write it in a more readable manner (as you could quickly overlook the negation in the condition). I guess it is a matter of taste, not runtime.

Comment (vhirtham) [reply weldx/asdf/util.py]: Jep, shouldn't have any effect on runtime. So your decision :wink

File: weldx/asdf/file.py (hunk 1 above)
Comment (vhirtham) [reply weldx/asdf/file.py:77]: Nothing in particular, just that in Python nothing is really protected from being manipulated. It was just a fun- comment. 😉

**Comment by marscher** [comment]
these "kernel died" errors during the notebook tests seem unrelated to the latest change.

File: weldx/asdf/file.py (hunk 1 above)
Comment (marscher) [reply weldx/asdf/file.py:77]: True, you can also re-define integers (try swapping one for zero is total fun).

**Comment by vhirtham** [comment]
> these "kernel died" errors during the notebook tests seem unrelated to the latest change.

Seems so, just restart the actions and see what happens

**Comment by marscher** [comment]
Now the number of hacks involved to get or set protected keys has been increased further (just for you @vhirtham :D)

File: weldx/asdf/util.py [hunk 3]
```diff
@@ -542,6 +553,82 @@ def _get_instance_shape(
     return None
 
 
+class _ProtectedViewDict(MutableMapping):
+    def __init__(self, protected_keys, data=None):
+        super(_ProtectedViewDict, self).__init__()
+        self.__data = data
```
Comment (CagtayFabry) [inline weldx/asdf/util.py:559]: nice try, but I'm pretty sure as long as something is called anything like `*data` @vhirtham will always find it 😉

**Pull request merged**